datafusion/physical-expr/src/intervals/interval_aritmetic.rs

```diff
@@ -1079,6 +1094,117 @@ mod tests {
         Interval::make(lower, upper, (false, false))
     }
 
+    #[test]
+    fn boolean_interval_test() -> Result<()> {
+        // Demonstrate / document how boolean Intervals work
+        #[derive(Debug)]
+        struct TestCase {
+            lower: bool,
+            upper: bool,
+            /// expected bounds when lower is open, upper is open
+            expected_open_open: (bool, bool),
+            /// expected bounds when lower is open, upper is closed
+            expected_open_closed: (bool, bool),
+            /// expected bounds when lower is closes, upper is open
+            expected_closed_open: (bool, bool),
+            // Not: closed/closed is the same as lower/upper
+        }
+
+        let cases = vec![
+            TestCase {
+                lower: false,
+                upper: false,
+                expected_open_open: (true, false), // whole range
+                expected_open_closed: (true, false),
+                expected_closed_open: (false, false),
+            },
+            TestCase {
+                lower: false,
+                upper: true,
+                expected_open_open: (true, false), // whole range
+                expected_open_closed: (true, true),
+                expected_closed_open: (false, false),
+            },
+            TestCase {
+                lower: true,
+                upper: false,
+                expected_open_open: (true, false), // whole range
+                expected_open_closed: (true, false),
+                expected_closed_open: (true, false),
+            },
+            TestCase {
+                lower: true,
+                upper: true,
+                expected_open_open: (true, false), // whole range
```

Comment:
An open `true` upper bound is mapped according to rule 2.

An open `true` lower bound is not mapped, right?

Reply:
In fact when I look at this test now it doesn't make sense as `(true, false)` isn't a valid `Interval` according to my understanding -- the interval should be `(false, true)` to represent the entire range 🤔

Reply:
A lower bound cannot be a true - open, and vice versa an upper bound cannot be a false - open. A boolean interval representing the entire range only can be [false, true].

Reply:
Thank you @berkaysynnada  -- this was my understanding as well. So given that this test passes with the current implementation, what do you suggest we do? 

1.  I wait for your refactoring PR
2. Fix the code to not create invalid intervals? 
3. Something else?

Reply:
I will be clarifying these issues and introducing a solid API in my PR. Is it possible to keep you waiting for a while, because they will all cause conflict for me.

Reply:
I'll handle any conflicts after your new PR lands -- no worries!